# Predicting incorrect student answers from DataShop data
<p style="margin:30px">
    <img width=50% src="https://www.featuretools.com/wp-content/uploads/2017/12/FeatureLabs-Logo-Tangerine-800.png" alt="Featuretools" />
</p>

In this tutorial, we show how to predict whether a student will succesfully answer a problem using a dataset from [CMU DataShop](https://pslcdatashop.web.cmu.edu/). While online courses are logistically efficient, the structure can make it more difficult for a teacher to understand how students are learning in their class. To try to fill in those gaps, we can apply machine learning. However, building an accurate machine learning model requires extracting information called **features**. Finding the right features is a crucial component of both finding a satisfactory answer and understanding what to do next. The process of **feature engineering** is made simple by [Featuretools](http://www.featuretools.com).

*If you're running this notebook yourself, please download the [geometry dataset](https://pslcdatashop.web.cmu.edu/DatasetInfo?datasetId=76) into the `data` folder in this repository. You will only need the `.txt` file. The infrastructure in this notebook will work with **any** DataShop dataset, but you will need to change the file name in the first cell.*

## Highlights
* Show how to import a DataShop dataset into featuretools
* Show efficacy of automatic feature generation with these datasets
* Show how to make custom primitives for stacking

# Step 1: Load data
At the beginning of any project, it is worthwhile to take a moment to think about how your dataset is structured.

In these datasets the unique events come from `transactions`: places where a student interacts with a system. Each transaction has a `time-index`, the time at which information in a row becomes known. Furthermore, the columns of those transactions have variables that can be grouped together. 

For instance, there are only 59 distinct students for the 6778 transactions we have in the geometry dataset. Those students log in to the system and have individual sessions. We can break down problems and problem steps in a similar way.

Featuretools stores data in an `EntitySet`. This is an abstraction which allows us to hold on to not only the data itself, but also to metadata like relationships and column types.

We create an entityset structure using the `datashop_to_entityset` function in [utils](utils.py). If you're interested in how `datashop_to_entityset` is structured, there's an associated notebook [entityset_function](entityset_function.ipynb) which explains choices made in more detail.

In [1]:
# Note that each branch is a one -> many relationship

# schools       students     problems
#        \        |         /
#   classes   sessions   problem steps
#          \     |       /
#           transactions  -- attempts
#

import utils

filename = 'data/ds2174_tx_All_Data_3991_2017_1128_123859.txt'
es = utils.datashop_to_entityset(filename)
es

Entityset: Dataset
  Entities:
    problem_steps (shape = [78, 50])
    sessions (shape = [59, 3])
    students (shape = [59, 2])
    transactions (shape = [6778, 26])
    problems (shape = [20, 2])
    ...And 3 more
  Relationships:
    transactions.Step Name -> problem_steps.Step Name
    problem_steps.Problem Name -> problems.Problem Name
    transactions.Session Id -> sessions.Session Id
    sessions.Anon Student Id -> students.Anon Student Id
    transactions.Class -> classes.Class
    ...and 2 more

Our `students` entity represents that: there are only 59 rows, one for each Anonymous student ID.

In [2]:
es['students'].df.head(3)

Anon Student Id  \
Anon Student Id                                                              
Stu_c0bf45c22dc46067350d304ce330067e  Stu_c0bf45c22dc46067350d304ce330067e   
Stu_af3a2f63bda8c1338556108cb8d519a0  Stu_af3a2f63bda8c1338556108cb8d519a0   
Stu_d7f18a5fa205a889b0c5b0b56a7127d3  Stu_d7f18a5fa205a889b0c5b0b56a7127d3   

                                     first_sessions_time  
Anon Student Id                                           
Stu_c0bf45c22dc46067350d304ce330067e 1996-02-01 00:00:00  
Stu_af3a2f63bda8c1338556108cb8d519a0 1996-02-01 00:00:02  
Stu_d7f18a5fa205a889b0c5b0b56a7127d3 1996-02-01 00:00:02

Featuretools allows us to make new entities as grouped by categorical values. Through this process of *normalization* we have created 8 connected entities from an initial table of transactions. We can look at what is left in `transactions` after normalization.

In [3]:
es['transactions'].head(3)

Sample Name  \
Transaction Id                                 
499a0a18d7b6d96d4ee9c16d4bead6f2    All Data   
d398b66148a76c537cba816efe946b85    All Data   
1c133061306fd8e099eb0e4f2ac21430    All Data   

                                                    Transaction Id  \
Transaction Id                                                       
499a0a18d7b6d96d4ee9c16d4bead6f2  499a0a18d7b6d96d4ee9c16d4bead6f2   
d398b66148a76c537cba816efe946b85  d398b66148a76c537cba816efe946b85   
1c133061306fd8e099eb0e4f2ac21430  1c133061306fd8e099eb0e4f2ac21430   

                                                      Session Id  \
Transaction Id                                                     
499a0a18d7b6d96d4ee9c16d4bead6f2  GEO-408d5ed7:10e14be5d3a:-8000   
d398b66148a76c537cba816efe946b85  GEO-408d5ed7:10e14be5d3a:-8000   
1c133061306fd8e099eb0e4f2ac21430  GEO-408d5ed7:10e14be5d3a:-6e40   

                                                Time   Time Zone  \
Transaction Id                                                     
499a0a18d7b6d96d4ee9c16d4bead6f2 1996-02-01 00:00:00  US/Eastern   
d398b66148a76c537cba816efe946b85 1996-02-01 00:00:01  US/Eastern   
1c133061306fd8e099eb0e4f2ac21430 1996-02-01 00:00:02  US/Eastern   

                                  Duration (sec) Student Response Type  \
Transaction Id                                                           
499a0a18d7b6d96d4ee9c16d4bead6f2               0               ATTEMPT   
d398b66148a76c537cba816efe946b85               1               ATTEMPT   
1c133061306fd8e099eb0e4f2ac21430               0               ATTEMPT   

                                  Student Response Subtype  \
Transaction Id                                               
499a0a18d7b6d96d4ee9c16d4bead6f2                       NaN   
d398b66148a76c537cba816efe946b85                       NaN   
1c133061306fd8e099eb0e4f2ac21430                       NaN   

                                 Tutor Response Type  Tutor Response Subtype  \
Transaction Id                                                                 
499a0a18d7b6d96d4ee9c16d4bead6f2              RESULT                     NaN   
d398b66148a76c537cba816efe946b85              RESULT                     NaN   
1c133061306fd8e099eb0e4f2ac21430              RESULT                     NaN   

                                         ...         Outcome  \
Transaction Id                           ...                   
499a0a18d7b6d96d4ee9c16d4bead6f2         ...               0   
d398b66148a76c537cba816efe946b85         ...               1   
1c133061306fd8e099eb0e4f2ac21430         ...               1   

                                                  Selection Action Input  \
Transaction Id                                                             
499a0a18d7b6d96d4ee9c16d4bead6f2  (CIRCLE-AREA_A QUESTION1)    NaN   NaN   
d398b66148a76c537cba816efe946b85  (CIRCLE-AREA_A QUESTION1)    NaN   NaN   
1c133061306fd8e099eb0e4f2ac21430           (AREA QUESTION1)    NaN   NaN   

                                  Feedback Text  Feedback Classification  \
Transaction Id                                                             
499a0a18d7b6d96d4ee9c16d4bead6f2            NaN                      NaN   
d398b66148a76c537cba816efe946b85            NaN                      NaN   
1c133061306fd8e099eb0e4f2ac21430            NaN                      NaN   

                                  Help Level Total Num Hints  Class  \
Transaction Id                                                        
499a0a18d7b6d96d4ee9c16d4bead6f2         NaN             NaN    NaN   
d398b66148a76c537cba816efe946b85         NaN             NaN    NaN   
1c133061306fd8e099eb0e4f2ac21430         NaN             NaN    NaN   

                                            End Time  
Transaction Id                                        
499a0a18d7b6d96d4ee9c16d4bead6f2 1996-02-01 00:00:00  
d398b66148a76c537cba816efe946b85 1996-02-01 00:00:02  
1c133061306fd8e09

# Step 2: Building Features
Next, we calculate a feature matrix on the `transactions` entity to try to predict the outcome of a given transaction. It's at this step that our previous setup pays off: we can automatically calculate features using data from the whole `EntitySet`. 

## Cutoff times
We are going to be generating features and doing predictive modeling on time-sensitive data. That comes with a high risk of label leakage. 

In this case, we are predicting if a student will get a particular problem correct. For a fixed problem, the feature "There exists an attempt number three" would be highly predictive of the result on attempts one and two. There can only be a third attempt if there first two attempts were wrong! In that way, storing future `attempt` information in a feature to predict `Outcome` would yield higher test accuracy than the model deserves. It's not ok to have the feature "There exists an attempt number three" while predicting attempts one and two because it contains information that can not be known at that point in time.

To circumvent that, we introduce the notion of [cutoff_times](https://docs.featuretools.com/automated_feature_engineering/handling_time.html). A `cutoff_time` has an index column and a datetime column indicating the last acceptable date we can use while generating features for a historical training example. We can also add in a label, which will be passed through Deep Feature Synthesis ([DFS](https://docs.featuretools.com/automated_feature_engineering/afe.html)) untouched so we can recover it later.

Setting cutoff times immediately mitigates the risk of fraudulently using future data, controls the number of predictions we make and controls what data is used while calculating features.


In [4]:
cutoff_times = es['transactions'].df[['Transaction Id', 'End Time', 'Outcome']]
cutoff_times.head()

Transaction Id  \
Transaction Id                                                       
499a0a18d7b6d96d4ee9c16d4bead6f2  499a0a18d7b6d96d4ee9c16d4bead6f2   
d398b66148a76c537cba816efe946b85  d398b66148a76c537cba816efe946b85   
1c133061306fd8e099eb0e4f2ac21430  1c133061306fd8e099eb0e4f2ac21430   
3ce42cc528ee77fb2db3b84c8b00ef39  3ce42cc528ee77fb2db3b84c8b00ef39   
82e2897fa00c8895bea39726e8dd303a  82e2897fa00c8895bea39726e8dd303a   

                                            End Time  Outcome  
Transaction Id                                                 
499a0a18d7b6d96d4ee9c16d4bead6f2 1996-02-01 00:00:00        0  
d398b66148a76c537cba816efe946b85 1996-02-01 00:00:02        1  
1c133061306fd8e099eb0e4f2ac21430 1996-02-01 00:00:02        1  
3ce42cc528ee77fb2db3b84c8b00ef39 1996-02-01 00:00:02        1  
82e2897fa00c8895bea39726e8dd303a 1996-02-01 00:00:02        0

With that in hand, we can guarentee that future values for `Outcome` won't be used for any calculations because we set the time index of `Outcome` to be after the cutoff time.

From there, we can call `ft.dfs` to generate our features and feature matrix. Deep Feature Synthesis creates features using reusable functions ([Primitives](https://docs.featuretools.com/automated_feature_engineering/primitives.html)). The algorithm attempts to combine primitives together with actual data to create a feature matrix. Here, we'll use the primitives `Sum`, `Mean`, `PercentTrue` and `Hour`. On an ordinary laptop, you should expect the following cell to take roughly 40 minutes to complete as there are more than 3000 unique cutoff times. For faster results, uncomment the approximate line. 

In [5]:
import featuretools as ft
from featuretools.primitives import Sum, Mean, Hour, PercentTrue
from featuretools.selection import remove_low_information_features

import pandas as pd
pd.options.display.max_columns = 500

fm, features = ft.dfs(entityset=es,
                      target_entity='transactions',
                      agg_primitives=[Sum, Mean, PercentTrue],
                      trans_primitives=[Hour],
                      max_depth=3,
                      # approximate='2m',
                      cutoff_time=cutoff_times[1000:],
                      verbose=True)

# Encode the feature matrix using One-Hot encoding
fm_enc, f_enc = ft.encode_features(fm, features)
fm_enc = fm_enc.fillna(0)
fm_enc = remove_low_information_features(fm_enc)

# Pop the label
label = fm_enc.pop('Outcome')

fm.tail()

Built 227 features
Elapsed: 37:51 | Remaining: 00:00 | Progress: 100%|██████████|| Calculated: 3068/3068 cutoff times


Class                     Selection  \
Transaction Id                                                          
a8899b5cab6774db36a200d174df1546    NaN       (SQUARE-AREA QUESTION2)   
2df16da47b60e3d15e098333b13ddf86    NaN       (SQUARE-AREA QUESTION3)   
ac8791497364a18b41957a72e6e1e6ec    NaN  (SCRAP-METAL-AREA QUESTION1)   
ef4433fd971485c7ffb3c9f15d9cfcf1    NaN  (SCRAP-METAL-AREA QUESTION2)   
b1ec2f9894ea80eeb7d01763976ba012    NaN  (SCRAP-METAL-AREA QUESTION3)   

                                  Attempt At Step Student Response Type  \
Transaction Id                                                            
a8899b5cab6774db36a200d174df1546                1               ATTEMPT   
2df16da47b60e3d15e098333b13ddf86                1               ATTEMPT   
ac8791497364a18b41957a72e6e1e6ec                1               ATTEMPT   
ef4433fd971485c7ffb3c9f15d9cfcf1                1               ATTEMPT   
b1ec2f9894ea80eeb7d01763976ba012                1               ATTEMPT   

                                  Tutor Response Subtype  Problem View  \
Transaction Id                                                           
a8899b5cab6774db36a200d174df1546                     NaN             2   
2df16da47b60e3d15e098333b13ddf86                     NaN             2   
ac8791497364a18b41957a72e6e1e6ec                     NaN             2   
ef4433fd971485c7ffb3c9f15d9cfcf1                     NaN             2   
b1ec2f9894ea80eeb7d01763976ba012                     NaN             2   

                                 Sample Name  Feedback Classification  \
Transaction Id                                                          
a8899b5cab6774db36a200d174df1546    All Data                      NaN   
2df16da47b60e3d15e098333b13ddf86    All Data                      NaN   
ac8791497364a18b41957a72e6e1e6ec    All Data                      NaN   
ef4433fd971485c7ffb3c9f15d9cfcf1    All Data                      NaN   
b1ec2f9894ea80eeb7d01763976ba012    All Data                      NaN   

                                 Level (Unit)  Feedback Text  \
Transaction Id                                                 
a8899b5cab6774db36a200d174df1546         Area            NaN   
2df16da47b60e3d15e098333b13ddf86         Area            NaN   
ac8791497364a18b41957a72e6e1e6ec         Area            NaN   
ef4433fd971485c7ffb3c9f15d9cfcf1         Area            NaN   
b1ec2f9894ea80eeb7d01763976ba012         Area            NaN   

                                                      Session Id  Input  \
Transaction Id                                                            
a8899b5cab6774db36a200d174df1546  GEO-408d5ed7:10e14be5d3a:-63b8    NaN   
2df16da47b60e3d15e098333b13ddf86  GEO-408d5ed7:10e14be5d3a:-63b8    NaN   
ac8791497364a18b41957a72e6e1e6ec  GEO-408d5ed7:10e14be5d3a:-63b8    NaN   
ef4433fd971485c7ffb3c9f15d9cfcf1  GEO-408d5ed7:10e14be5d3a:-63b8    NaN   
b1ec2f9894ea80eeb7d01763976ba012  GEO-408d5ed7:10e14be5d3a:-63b8    NaN   

                                  Action  Student Response Subtype  \
Transaction Id                                                       
a8899b5cab6774db36a200d174df1546     NaN                       NaN   
2df16da47b60e3d15e098333b13ddf86     NaN                       NaN   
ac8791497364a18b41957a72e6e1e6ec     NaN                       NaN   
ef4433fd971485c7ffb3c9f15d9cfcf1     NaN                       NaN   
b1ec2f9894ea80eeb7d01763976ba012     NaN                       NaN   

                                 Tutor Response Type   Time Zone  \
Transaction Id                                                     
a8899b5cab6774db36a200d174df1546              RESULT  US/Eastern   
2df16da47b60e3d15e098333b13ddf86              RESULT  US/Eastern   
ac8791497364a18b41957a72e6e1e6ec              RESULT  US/Eastern   
ef4433fd971485c7ffb3c9f15d9cfcf1              RESULT  US/Eastern   
b1ec2f9894ea80eeb7d01763976ba012              RESULT  US/Eastern   

                      

Above, you can scroll to the right to see the 227 features we created. If you look at the column names, you can see that we've done more than individually apply primitives one at a time to the raw data. Features were stacked and combined across entities in an exhaustive way. Using Deep Feature Synthesis is powerful because it greatly increases the likelihood of finding important features while decreasing the workload of the data scientist.

# Step 3: Making predictions

Here we split the data into two parts using `train_test_split` from scikit-learn. Notice that we don't want the splitter to shuffle our data, since that has the risk to leak labels in time sensitive data. 

We can do feature selection with [Recursive Feature Elimination](http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html). This recursively removes features by checking feature importances (according to some model) with smaller and smaller feature sets. Here we'll set `RFE` to select 20 features.  

In [6]:
from sklearn.model_selection import TimeSeriesSplit, train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score
from utils import feature_importances
from sklearn.feature_selection import RFE

# 1. Split X and y into a train and test set
X_train, X_test, y_train, y_test = train_test_split(fm_enc, label, shuffle=False)

# 2. Select features using RFE
clf = RandomForestClassifier()
estimator = clf
selector = RFE(estimator, 20, step=1)
selector = selector.fit(X_train, y_train)
X_train.iloc[:, selector.support_].tail()

sessions.PERCENT_TRUE(transactions.Outcome)  \
Transaction Id                                                                  
1cdbd9f38140e09dc7fa4e642c0b44f1                                     0.744681   
387f37198680f1e9e6d449a9c62ef8d3                                     0.767347   
aa77623c3440cb105a939a1aca1d4fe2                                     0.753846   
f4812c4adc51484bf1cdfc3be42a87c0                                     0.820652   
ff861aa2fd87d304cb0759526cc1ed84                                     0.744856   

                                  problem_steps.SUM(transactions.Duration (sec))  \
Transaction Id                                                                     
1cdbd9f38140e09dc7fa4e642c0b44f1                                          1290.0   
387f37198680f1e9e6d449a9c62ef8d3                                           477.0   
aa77623c3440cb105a939a1aca1d4fe2                                          4291.0   
f4812c4adc51484bf1cdfc3be42a87c0                                           453.0   
ff861aa2fd87d304cb0759526cc1ed84                                           565.0   

                                  sessions.MEAN(transactions.Is Last Attempt)  \
Transaction Id                                                                  
1cdbd9f38140e09dc7fa4e642c0b44f1                                     0.359712   
387f37198680f1e9e6d449a9c62ef8d3                                     0.360656   
aa77623c3440cb105a939a1aca1d4fe2                                     0.437500   
f4812c4adc51484bf1cdfc3be42a87c0                                     0.502732   
ff861aa2fd87d304cb0759526cc1ed84                                     0.344398   

                                  problem_steps.PERCENT_TRUE(transactions.Outcome)  \
Transaction Id                                                                       
1cdbd9f38140e09dc7fa4e642c0b44f1                                          0.832370   
387f37198680f1e9e6d449a9c62ef8d3                                          0.609375   
aa77623c3440cb105a939a1aca1d4fe2                                          0.765182   
f4812c4adc51484bf1cdfc3be42a87c0                                          0.837209   
ff861aa2fd87d304cb0759526cc1ed84                                          0.822222   

                                  sessions.MEAN(transactions.Problem View)  \
Transaction Id                                                               
1cdbd9f38140e09dc7fa4e642c0b44f1                                  1.787234   
387f37198680f1e9e6d449a9c62ef8d3                                  1.440816   
aa77623c3440cb105a939a1aca1d4fe2                                  1.907692   
f4812c4adc51484bf1cdfc3be42a87c0                                  1.929348   
ff861aa2fd87d304cb0759526cc1ed84                                  1.205761   

                                  problem_steps.MEAN(transactions.Duration (sec))  \
Transaction Id                                                                      
1cdbd9f38140e09dc7fa4e642c0b44f1                                         7.500000   
387f37198680f1e9e6d449a9c62ef8d3                                         7.571429   
aa77623c3440cb105a939a1aca1d4fe2                                         4.351927   
f4812c4adc51484bf1cdfc3be42a87c0                                        10.785714   
ff861aa2fd87d304cb0759526cc1ed84                                        13.139535   

                                  attempts.MEAN(transactions.Duration (sec))  \
Transaction Id                                                                 
1cdbd9f38140e09dc7fa4e642c0b44f1                                   12.038432   
387f37198680f1e9e6d449a9c62ef8d3                                   12.038432   
aa77623c3440cb105a939a1aca1d4fe2                                    3.707234   
f4812c4adc51484bf1cdfc3be42a87c0                                   12.036391   
ff861aa2fd87d304cb0759526cc1ed84                                    3.704932   

  

Finally, we can train a Random Forest Classifier to make predictions. Those predictions can be checked against our `y_test` from above, and scored with a [roc_auc_score](https://en.wikipedia.org/wiki/Receiver_operating_characteristic). Below we'll train and score our model and output the five most important features according to this model.

In [7]:
# 3. Train a Random Forest Classifier
clf.fit(selector.transform(X_train), y_train)

# 4. Make predictions and score
preds = clf.predict(selector.transform(X_test))
print("Auc score of {:.3f}".format(roc_auc_score(y_test, preds)))

feats = feature_importances(X_train.iloc[:, selector.support_], clf)

Auc score of 0.555
Feature Importances: 
1: attempts.PERCENT_TRUE(transactions.Outcome)
2: problem_steps.PERCENT_TRUE(transactions.Outcome)
3: attempts.MEAN(transactions.Duration (sec))
4: problem_steps.problems.PERCENT_TRUE(transactions.Outcome)
5: attempts.SUM(transactions.Duration (sec))
-----



Let's examine a feature. The feature `problem_steps.MEAN(transactions.Duration (sec))` is the average time spent on a given problem step. It's easy to see how 'amount of time people spend on this problem' might be related to problem difficulty and ultimately the `Outcome` of a given attempt.

# Next Steps
This notebook showed how to structure your data and make predictions with machine learning. Rather than spending time creating features, it's now possible to explore the relationships and implications betweem thousands of features directly. Reasonable next steps might be to:
1. Try [plotting](#Appendix:-Plotting) some of the generated features
2. Run feature selection and tune the machine learning model
3. Explore other prediction problems on this `EntitySet`




# Appendix: Plotting
Here, we'll look at a couple of important features as created above. We can use plots to help us understand why certain automatically generated features are good. Here, we'll plot two important features from the model above and match results from the model to our own intuition.

In [41]:
from bokeh.io import show, output_notebook, output_file

output_notebook()
output_file('difficulty_vs_time.html')

p = utils.datashop_plot(fm,
                        col1='problem_steps.problems.PERCENT_TRUE(transactions.Outcome)',
                        col2='problem_steps.problems.MEAN(transactions.Duration (sec))',
                        label=label,
                        names=['Problem difficulty versus problem time', 
                               'Success rate on this problem', 
                               'Average time on this problem'])
show(p)


Loading BokehJS ...

![](data/images/exampleimage.png)

If you're interested in understanding particular points and clusters, we have saved an html interactive version which you can download [here](data/images/difficulty_vs_time.html). That version will allow you to zoom in and hover over individual points to see which problem step and problem it is. 

Notice that while a feature like *Success rate on this problem* might only have one value if we use all of the data, the graph here shows that data changing with time. To start our analysis, let's get a baseline for the data. The blue dots represent a successful answer while the grey dots indicate an incorrect answer. We can ask how often students are correct on average:

In [13]:
print('Overall success rate is {:.2f}%'.format(100*np.mean(fm['Outcome'])))

Overall success rate is 0.79


That is, if you were to pick a point at random, there's a roughly 75% chance it will be a correct answer. There are sections of this graph where that sample is more likely to be correct, and more likely to be incorrect, which can be picked up by the decision trees that make up our Random Forest. From the graph it looks like there is a spike of correct answers in this dataset near problems that take 10 seconds. We can verify that directly:

In [36]:
maxtime = 15

print('If problem takes more than {} seconds: {:.2f}% of problems answered correctly'.format(maxtime,
    100 * np.mean(fm[
        (fm['problem_steps.problems.MEAN(transactions.Duration (sec))']>=maxtime)]['Outcome'])))

If problem takes more than 15 seconds: 74.80% of problems answered correctly


In other words, the average time spent on a problem is an indicator of whether or not a student will answer the problem correctly in this dataset. There are a number of possible interpretations and testable hypotheses associated to that. It is clear that the averages don't tell the whole story of what's going on. Let's look at success rate as sorted by problem.



In [35]:
split_line = .85

print('If Success Rate > {}: {:.2f}% of problems answered correctly'.format(split_line,
    100 * np.mean(fm[
        (fm['problem_steps.problems.PERCENT_TRUE(transactions.Outcome)']>=split_line)]['Outcome'])))
print('Problems with higher success than {}: {}'.format(split_line, 
                                                        fm[fm['problem_steps.problems.PERCENT_TRUE(transactions.Outcome)']
                                                           >=split_line]['problem_steps.Problem Name'].unique()))


If Success Rate > 0.85: 89.24% of problems answered correctly
Problems with higher success than 0.85: ['BUILDING_A_SIDEWALK' 'POGS' 'PAINTING_THE_WALL' 'CIRCLE_O'
 'DESIGNING_A_QUILT' 'DOG_ON_A_ROPE']


That is, of the 20 problems in this data set, only 6 have a success rate that was over 85%. In that way the machine learning has indicated that how previous students have done on the problem is a good predictor of how they will do inside of this dataset.

In addition to our earlier conclusion that "the problems that took a long time had worse scores", we have a secondary conclusion that "some problems are harder than others". What makes this line of inquiry interesting is that we didn't have to do very much work to reveal hard questions. In that way we have used automated feature engineering to make explicit our implicit understanding of this dataset.

# Feature Labs
If you have questions, comments or concerns regarding the notebook feel free to [reach out](https://stackoverflow.com/questions/tagged/featuretools).